File: cpp/tests/back_smoke.cpp
// Regression test for Back after an in-page anchor. History used to hold
// documents only, so following "#section" and pressing Back reopened the
// PREVIOUS DOCUMENT instead of returning to where the link was read. Drives the
// real MainWindow: scroll, click an anchor, press Back, read the offset back.
// Run with QT_QPA_PLATFORM=offscreen.

#include "mainwindow.h"

#include <QApplication>
#include <QFile>
#include <QTemporaryDir>
#include <QTimer>
#include <QVariant>
#include <QWebEnginePage>
#include <QWebEngineView>

#include <cstdio>
#include <cstdlib>

namespace {

MainWindow *window = nullptr;
double before = 0;
double afterJump = 0;

void js(const QString &code, const std::function<void(const QVariant &)> &then) {
    window->webView()->page()->runJavaScript(code, then);
}

void fail(const char *why) {
    std::printf("BACK: FAIL (%s)\n", why);
    std::exit(1);
}

} // namespace

int main(int argc, char **argv) {
    QApplication app(argc, argv);

    QTemporaryDir dir;
    // The link sits well down the document and its target further still, so a
    // successful jump and a successful return are both unambiguous.
    QString filler;
    for (int i = 0; i < 120; ++i)
        filler += "filler\n\n";
    QString tail;
    for (int i = 0; i < 200; ++i)
        tail += "filler\n\n";
    QFile f(dir.filePath("doc.md"));
    if (!f.open(QIODevice::WriteOnly | QIODevice::Text))
        fail("could not write doc.md");
    f.write(("# Top\n\n" + filler + "[jump](#the-target)\n\n" + tail + "## The Target\n\ncontent\n")
                .toUtf8());
    f.close();

    MainWindow w;
    window = &w;
    w.resize(900, 720);
    w.show();
    w.openPath(dir.filePath("doc.md"));

    QObject::connect(w.webView(), &QWebEngineView::loadFinished, [&](bool ok) {
        if (!ok)
            fail("the document did not load");
        QTimer::singleShot(600, [] {
            js("document.scrollingElement.scrollTop = 3000;"
               " document.scrollingElement.scrollTop",
               [](const QVariant &v) {
                   before = v.toDouble();
                   js("document.querySelector('a[href=\"#the-target\"]').click(); 1",
                      [](const QVariant &) {
                          QTimer::singleShot(600, [] {
                              js("document.scrollingElement.scrollTop", [](const QVariant &v) {
                                  afterJump = v.toDouble();
                                  if (afterJump <= before)
                                      fail("the anchor did not scroll anywhere");
                                  window->goBack();
                                  QTimer::singleShot(700, [] {
                                      js("document.scrollingElement.scrollTop",
                                         [](const QVariant &v) {
                                             const double back = v.toDouble();
                                             const bool ok = qAbs(back - before) < 5;
                                             std::printf("BACK: %s (before %.0f, jump %.0f, "
                                                         "back %.0f)\n",
                                                         ok ? "PASS" : "FAIL", before, afterJump,
                                                         back);
                                             std::exit(ok ? 0 : 1);
                                         });
                                  });
                              });
                          });
                      });
               });
        });
    });

    QTimer::singleShot(25000, [] { fail("timed out"); });
    return app.exec();
}

File: python/tests/back_smoke.py
"""Back after an in-page anchor must return to where the link was read.

History used to hold documents only, so following "#section" and pressing Back
reopened the previous *document* instead of coming back to the position the link
was on. Drives the real MainWindow offscreen: scroll, click the anchor, press
Back, read the offset back out. Exit 0 = pass.
"""

from __future__ import annotations

import sys
import tempfile
from pathlib import Path

from PySide6.QtCore import QTimer
from PySide6.QtWidgets import QApplication

from marklens.app import MainWindow

# The link sits well down the document and its target further still, so both a
# successful jump and a successful return are unambiguous.
DOC = (
    "# Top\n\n"
    + "filler\n\n" * 120
    + "[jump](#the-target)\n\n"
    + "filler\n\n" * 200
    + "## The Target\n\ncontent\n"
)


def main() -> int:
    folder = Path(tempfile.mkdtemp())
    doc = folder / "doc.md"
    doc.write_text(DOC, encoding="utf-8")

    app = QApplication(sys.argv)
    window = MainWindow()
    window.resize(900, 720)
    window.show()
    window.open_path(doc)

    state: dict[str, object] = {}

    def js(code: str, then) -> None:
        window._view.page().runJavaScript(code, then)

    def scrolled_down(value: object) -> None:
        state["before"] = float(value or 0)
        js("document.querySelector('a[href=\"#the-target\"]').click(); 1",
           lambda _: QTimer.singleShot(600, read_jump))

    def read_jump() -> None:
        js("document.scrollingElement.scrollTop", after_jump)

    def after_jump(value: object) -> None:
        state["jump"] = float(value or 0)
        state["back_enabled"] = window._back_action.isEnabled()
        window._go_back()
        QTimer.singleShot(700, lambda: js("document.scrollingElement.scrollTop", after_back))

    def after_back(value: object) -> None:
        state["back"] = float(value or 0)
        app.quit()

    def start(ok: bool) -> None:
        if not ok:
            print("BACK: FAIL (document did not load)")
            app.quit()
            return
        QTimer.singleShot(
            600,
            lambda: js(
                "document.scrollingElement.scrollTop = 3000;"
                " document.scrollingElement.scrollTop",
                scrolled_down,
            ),
        )

    window._view.loadFinished.connect(start)
    QTimer.singleShot(30000, app.quit)
    app.exec()

    before = float(state.get("before", 0) or 0)
    jump = float(state.get("jump", 0) or 0)
    back = float(state.get("back", -1) or -1)
    ok = jump > before and abs(back - before) < 5 and state.get("back_enabled") is True
    print(
        f"BACK: {'PASS' if ok else 'FAIL'} (before {before:.0f}, jump {jump:.0f}, "
        f"back {back:.0f}, button {state.get('back_enabled')})"
    )
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
